base/src/main/java/org/aya/resolve/visitor/ExprResolver.java

```diff
@@ -29,112 +30,167 @@
 /**
  * Resolves bindings.
  *
- * @param allowedGeneralizes will be filled with generalized vars if allowGeneralized,
- *                           and represents the allowed generalized level vars otherwise
- * @author re-xyr, ice1000
+ * @author re-xyr, ice1000, kiva
  * @implSpec allowedGeneralizes must be linked map
  * @see StmtResolver
  */
 public record ExprResolver(
-  @NotNull Options options,
-  @NotNull MutableMap<GeneralizedVar, Expr.Param> allowedGeneralizes,
-  @NotNull MutableList<TyckOrder> reference,
-  @NotNull MutableStack<Where> where,
-  @Nullable Consumer<TyckUnit> parentAdd
-) implements Expr.Visitor<Context, @NotNull Expr> {
-  @Override public @NotNull Expr visitRef(Expr.@NotNull RefExpr expr, Context p) {
-    return expr;
-  }
-
-  public @NotNull ImmutableSeq<Expr.@NotNull Param> visitParams(@NotNull ImmutableSeq<Expr.@NotNull Param> params, Context p) {
-    return params.map(param -> {
-      var oldType = param.type();
-      var type = oldType.accept(this, p);
-      if (type == oldType) return param;
-      return new Expr.Param(param, type);
-    });
-  }
-
-  @Override public @NotNull Expr visitError(Expr.@NotNull ErrorExpr error, Context p) {
-    return error;
-  }
-
-  @Override public @NotNull Expr visitRawUniv(Expr.@NotNull RawUnivExpr expr, Context p) {
-    return expr;
-  }
-
-  @Override public @NotNull Expr visitMetaPat(Expr.@NotNull MetaPat metaPat, Context p) {
-    return metaPat;
-  }
+  @Override @NotNull ExprView view,
+  @NotNull State state,
+  @NotNull Context context
+) implements ExprOps {
```

Comment:
Does it have to be an implementation of `ExprOps`?

Reply:
I think probably no. And I even think that it's not necessary to rewrite the expr resolver with ExprView (at least right now):

Reply:
But [this commit](https://github.com/aya-prover/aya-dev/pull/374/commits/bdbce8ceb33f8035747d99d39958ca0b437cfba9) is still useful and I would like to give it a cherry-pick 🍒 .

Reply:
> But [this commit](https://github.com/aya-prover/aya-dev/pull/374/commits/bdbce8ceb33f8035747d99d39958ca0b437cfba9) is still useful and I would like to give it a cherry-pick 🍒 .

Of course, I would like to merge the first commit and the part that you mark yourself as an author

Reply:
bors d=imkiva